persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java

```diff
@@ -274,32 +287,43 @@ public ConnectionFactory getConnectionFactory() {
                ps = conn.prepareStatement(sql);
                ps.setLong(1, timeService.wallClockTime());
                rs = ps.executeQuery();
+               int batchSize = configuration.maxBatchSize();
+               List<PossibleExpirationNotification> list;
+
+               if (key2StringMapper instanceof TwoWayKey2StringMapper) {
+                  list = new ArrayList<>(batchSize);
+               } else {
+                  list = null;
+                  PERSISTENCE.twoWayKey2StringMapperIsMissing(TwoWayKey2StringMapper.class.getSimpleName());
+               }
+               long purgedAmount = 0;
 
-               try (PreparedStatement batchDelete = conn.prepareStatement(tableManager.getDeleteRowSql())) {
-                  int affectedRows = 0;
-                  boolean twoWayMapperExists = key2StringMapper instanceof TwoWayKey2StringMapper;
+               try (PreparedStatement batchDelete = conn.prepareStatement(tableManager.getDeleteRowWithExpirationSql())) {
+                  long possibleAmount = 0;
                   while (rs.next()) {
-                     affectedRows++;
                      String keyStr = rs.getString(2);
                      batchDelete.setString(1, keyStr);
+                     long expiryTime = rs.getLong(3);
+                     batchDelete.setLong(2, expiryTime);
                      batchDelete.addBatch();
 
-                     if (twoWayMapperExists) {
+                     if (list != null) {
                         InputStream inputStream = rs.getBinaryStream(1);
-                        Object key = ((TwoWayKey2StringMapper) key2StringMapper).getKeyMapping(keyStr);
-                        MarshalledValue value = unmarshall(inputStream, marshaller);
-                        unicastProcessor.onNext(marshalledEntryFactory.create(key, value));
+                        list.add(new PossibleExpirationNotification(keyStr, inputStream));
```

Comment:
I don't think it's safe to keep the InputStream after moving to the next row. The `getBinaryStream()` javadoc says

> Note: All the data in the returned stream must be read prior to getting the value of any other column. The next call to a getter method implicitly closes the stream. Also, a stream may return 0 when the method InputStream.available is called whether there is data available or not

Reply:
Good catch. I was trying to avoid unmarshalling the MarshalledValue, but we need to keep it per the javadoc.